lib/state/data/reducer.ts

```diff
@@ -553,3 +543,56 @@ export default combineReducers({
   preferences,
   tags,
 });
+
+const crossSliceReducer = (
+  state: DataState,
+  action: A.ActionType
+): DataState => {
+  switch (action.type) {
+    case 'RESTORE_NOTE_REVISION': {
+      const revision = state.noteRevisions
+        .get(action.noteId)
+        ?.get(action.version);
+
+      if (!revision) {
+        return state;
+      }
+
+      const note = state.notes.get(action.noteId);
+      const noteEmailTags =
+        note?.tags.filter((tagName) => isEmailTag(tagName)) ?? [];
+
+      const revisionCanonicalTags = revision.tags.filter((tagName) => {
+        const tagHash = t(tagName);
+        const hasTag = state.tags.has(tagHash);
+        return !isEmailTag(tagName) && (hasTag || action.includeDeletedTags);
+      });
+
+      const notesMap = new Map(state.notes).set(action.noteId, {
+        ...revision,
+        tags: [...noteEmailTags, ...revisionCanonicalTags],
+      });
+
+      // Restore potential deleted tags
+      const tagsMap = new Map(state.tags);
+      revisionCanonicalTags.forEach((tagName) => {
+        const tagHash = t(tagName);
+        if (!tagsMap.has(tagHash)) {
+          tagsMap.set(tagHash, { name: tagName });
+        }
+      });
+
+      return { ...state, notes: notesMap, tags: tagsMap };
+    }
+    default:
+      return state;
+  }
+};
+
+const dataReducer: A.Reducer<DataState> = (state, action) => {
+  const intermediateState = combinedReducer(state, action);
+  const finalState = crossSliceReducer(intermediateState, action);
+  return finalState;
+};
```

Comment:
We would have to be careful with this approach because, for the cross slice reducer, we use the resulting state from `combineReducer`.

If for example we watch for the `RESTORE_NOTE_REVISION` action in one of the slices of `combineReducer` and modify the state, we could have an undesired behaviour when executing the cross slice reducer.